docs/dev/plugins.md

```diff
@@ -328,6 +328,62 @@ no_command do
 end
 ```
 
+## Implementing Reporter Plugins
+
+Reporter plugins offer the opportunity to customize or create entirely new output formats for Chef InSpec. Reporter plugins operate at the very end of the Chef InSpec run when all test data has finalized.
+
+### Declare your plugin activators
+
+In your `plugin.rb`, include one or more `reporter` activation blocks. The activation block name will be matched against the value passed into the `--reporter` option. If a match occurs, your activator will fire, which loads any needed libraries, and return your implementation class.
+
+#### Reporter Activator Example
+
+```ruby
+
+# In plugin.rb
+module InspecPlugins::Sweeten
+  class Plugin < Inspec.plugin(2)
+    # ... other plugin stuff
+
+    reporter :sweet do
+      require_relative 'reporter.rb'
+      InspecPlugins::Sweeten::Reporter
+    end
+  end
+end
+```
+
+Like any activator, the block above will only be called if needed. For Reporter plugins, the plugin system examines the `--reporter` argument, or the `reporter:` JSON config option, and looks for the activation name as a prefix. Multiple Reporter activations may occur if several different names match, though each activation will only occur once.
+
+```bash
+you@machine $ inspec exec --reporter sweet # Your Reporter implementation is activated and executed
+you@machine $ inspec exec --reporter json  # Your Reporter implementation is not activated
+```
+
+### Implementation class for Reporters
+
+In your `reporter.rb`, you should begin by requesting the superclass from `Inspec.plugin`:
+
+```ruby
+module InspecPlugins::Sweeten
+  class Reporter < Inspec.plugin(2, :reporter)
+    def render
+      # examine run_data and call output()
+    end
+  end
+end
+```
+
+### Implementing your Reporter
+
+#### Implement render()
+
+The primary responsibility you must fulfill is to implement render. Typically, you will examine the `run_data` Hash, which is provided as an accessor. Call `output(String, newline_wanted = true)` to send output.
+
+#### The run_data structure
+
+The `run_data` object contains all data from the Chef InSpec run. This object is a Hash, but includes many fields and is often large. No specific documentation exists for the `run_data` object. See [the legacy JSON reporter](https://github.com/inspec/inspec/blob/2e887a94afcca819da781d4774aa2a5a0b56785e/lib/inspec/reporters/json.rb#L10) for one example of how to iterate over the object.
```

Comment:
If this is intended as a user-facing feature, one worry here is that this effectively makes `run_data` and every object reachable from `run_data` part of the public facing API. Are we OK with that? I took a peak of the surface area there and it doesn't seem _too_ bad on the surface, but my guess is that exposing `Inspec::Profile` probably means most of our major data structures are somehow reachable from run_data.

To completely fix the problem probably involves boiling the ocean, but I'm wondering if we want to do a few things:

1) Wrap the run_data hash in a class with accessors before passing it to reporters. 
2) Extend some of this document with the core bits of this structure that we think we can keep stable over time.
2) Add a warning that accessing data other than those documented might result in their reporter breaking in the future?

(note, this is an actual question, I yield to y'alls judgement)

Reply:
Are you concerned that, say for example, someone builds a very involved/intricate reporter plugin that expects `run_data` to be a certain shape and format? And then if we modify it at any point in the future we risk breaking the reporter?

Reply:
@Schwad Yup

Reply:
This is a good point @stevendanna 
I think the more general problem is that there is no API into InSpec for plugins to use.

I believe that currently plugin maintainers are responsible for keeping their plugins working if they want to change InSpec version.
That is a. Unfriendly and b. Possibly very risky when it comes to reporting - their reporter plugin may continue to complete but start to miss bits of result data.

Another option is to maintain a versioned schema for run_data's structure and have reporter plugins need to declare what schema version they are intended for. It will still need plugin developers to keep up to date with InSpec changes, but at least their runs will outright fail instead of resulting in bad report data

Reply:
I'm put some time into making an object model for run_data and documenting as I go. I'm keeping it backwards compatible with Hash access so the existing reports should still work with it.